File: example/xplayctl.go
package main

// 20220520

import (
	"bufio"
	"encoding/json"
	"flag"
	"fmt"
	"log"
	"net"
	"net/http"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"
)

// Sync Wait Group
var wg sync.WaitGroup

// Connect Address
var addr = flag.String("addr", "127.0.0.1:8700", "xplay tcp address")

// Query
var query = flag.Bool("query", false, "query playing status")

// Snap
var snap = flag.Bool("snap", false, "screen snapshot")

// Move
var move = flag.Bool("move", false, "move playing")

// Change
var change = flag.Bool("change", false, "change playing")

// General
// var id = flag.String("id", "", "debug id") // 自动生成
var start = flag.Int64("start", -1, "start time(ms)")

// Stop
var stop = flag.Bool("stop", false, "stop or play")
var all = flag.Bool("all", false, "stop all")
var ids = flag.String("ids", "", "stop ids")

// Play
var play = flag.Bool("play", false, "play or stop")
var libName = flag.String("libName", "", "video、pic、sequence、camera、gif、qrcode、text、scroll、background")

// Play Params
var zIndex = flag.Int("zIndex", 10, "1-999")
var rect = flag.String("rect", "0,0,1920,1080", "")
var screen_mode = flag.String("screen_mode", "landscape", "landscape、portrait")
var screen_rotate = flag.Int("screen_rotate", 0, "landscape: 0、180 or portrait: 90、270")
var path = flag.String("path", "", "file path")
var content = flag.String("content", "", "data content")
var offset = flag.Int64("offset", -1, "video offset(ms)")
var timeout = flag.Int64("timeout", -1, "video stream timeout(ms)")

// Change Params
var newIndex = flag.Int("newIndex", 0, "1-999")

// Text
var font_ttf = flag.String("font_ttf", "", "TrueTypeFont")
var font_size = flag.Int("font_size", 18, "")
var color = flag.String("color", "rgba(0,128,0,100%)", "")
var bgcolor = flag.String("bgcolor", "rgba(0,0,0,20%)", "")
var align = flag.String("align", "center", "center、right、left")
var style = flag.String("style", "normal", "normal、bold、italic、underline、strikethrough")

// Camera
var device = flag.String("device", "/dev/video0", "camera device")
var camera_width = flag.Int("camera_width", 1280, "camera video width")
var camera_height = flag.Int("camera_height", 720, "camera video height")

// Scroll
var speed = flag.Int("speed", 1, "move pixel / frame")
var orientation = flag.String("orientation", "horizontal", "horizontal | vertical")

// Toast
var toast_type = flag.String("toast_type", "notice", "notice、success、warning、error")
var duration = flag.Int("duration", 0, "toast duration(s)")

type XPlay struct {
	conn *net.TCPConn
}

func NewXPlay() *XPlay {
	xplay := new(XPlay)
	xplay.conn = nil
	return xplay
}

// 连接 XPlay
func (this *XPlay) connect(_addr string) error {
	addr, err := net.ResolveTCPAddr("tcp", _addr)
	if err != nil {
		return err
	}
	conn, err := net.DialTCP("tcp", nil, addr)
	if err != nil {
		return err
	}
	this.conn = conn
	return nil
}

// 从 XPlay 接收返回
func (this *XPlay) result() {
	fmt.Println(">>> RESULT >>>")
	br := bufio.NewReader(this.conn)
	for {
		line, _, err := br.ReadLine()
		if err != nil {
			log.Println(err.Error())
			break
		}
		data := string(line)
		fmt.Println(data)
		if strings.Trim(data, " ") == "#End" {
			break
		}
	}
}

// 发送指令到 XPlay
func (this *XPlay) send(_data map[string]interface{}) error {
	js, _ := json.MarshalIndent(_data, "", "  ")
	fmt.Println(">>> SEND >>>")
	fmt.Println(string(js))
	_, err := this.conn.Write(append(js, []byte("\n#End\n")...))
	return err
}

// 停止全部
func (this *XPlay) stop_all() error {
	data := make(map[string]interface{})
	data["id"] = fmt.Sprintf("STOP_%d", time.Now().Unix())
	data["type"] = "stop"
	data["start"] = *start
	data["params"] = map[string]bool{"all": true}
	return this.send(data)
}

// 停止指定层
func (this *XPlay) stop(_ids string) error {
	vs := strings.Split(_ids, ",")
	data := make(map[string]interface{})
	params := make(map[string]interface{})
	params["ids"] = vs
	data["id"] = fmt.Sprintf("STOP_%d", time.Now().Unix())
	data["type"] = "stop"
	data["start"] = *start
	data["params"] = params
	return this.send(data)
}

func (this *XPlay) play() error {
	params := make(map[string]interface{})
	rs := strings.Split(*rect, ",")
	x, _ := strconv.Atoi(rs[0])
	y, _ := strconv.Atoi(rs[1])
	width, _ := strconv.Atoi(rs[2])
	height, _ := strconv.Atoi(rs[3])
	if (*libName) == "sequence" {
		params["zIndex"] = *zIndex
		params["left"] = x
		params["top"] = y
		params["width"] = width
		params["height"] = height
		params["screen_mode"] = *screen_mode
		params["screen_rotate"] = *screen_rotate
		params["path"] = this.path()
		params["duration"] = *duration
	} else if (*libName) == "video" {
		params["zIndex"] = *zIndex
		params["left"] = x
		params["top"] = y
		params["width"] = width
		params["height"] = height
		params["screen_mode"] = *screen_mode
		params["screen_rotate"] = *screen_rotate
		params["path"] = this.path()
		params["offset"] = *offset
		params["timeout"] = *timeout
	} else if (*libName) == "pic" {
		params["zIndex"] = *zIndex
		params["left"] = x
		params["top"] = y
		params["width"] = width
		params["height"] = height
		params["screen_mode"] = *screen_mode
		params["screen_rotate"] = *screen_rotate
		params["path"] = this.path()
	} else if (*libName) == "gif" {
		params["zIndex"] = *zIndex
		params["left"] = x
		params["top"] = y
		params["width"] = width
		params["height"] = height
		params["screen_mode"] = *screen_mode
		params["screen_rotate"] = *screen_rotate
		params["path"] = this.path()
	} else if (*libName) == "qrcode" {
		params["zIndex"] = *zIndex
		params["left"] = x
		params["top"] = y
		params["width"] = width
		params["height"] = height
		params["screen_mode"] = *screen_mode
		params["screen_rotate"] = *screen_rotate
		params["content"] = *content
	} else if (*libName) == "camera" {
		params["zIndex"] = *zIndex
		params["left"] = x
		params["top"] = y
		params["width"] = width
		params["height"] = height
		params["screen_mode"] = *screen_mode
		params["screen_rotate"] = *screen_rotate
		params["device"] = *device
		params["camera_width"] = *camera_width
		params["camera_height"] = *camera_height
	} else if (*libName) == "text" {
		params["zIndex"] = *zIndex
		params["left"] = x
		params["top"] = y
		params["width"] = width
		params["height"] = height
		params["screen_mode"] = *screen_mode
		params["screen_rotate"] = *screen_rotate
		params["content"] = *content
		params["font_ttf"] = *font_ttf
		params["font_size"] = *font_size
		params["color"] = *color
		params["bgcolor"] = *bgcolor
		params["align"] = *align
		params["style"] = *style
	} else if (*libName) == "scroll" {
		params["zIndex"] = *zIndex
		params["left"] = x
		params["top"] = y
		params["width"] = width
		params["height"] = height
		params["screen_mode"] = *screen_mode
		params["screen_rotate"] = *screen_rotate
		params["content"] = *content
		params["font_ttf"] = *font_ttf
		params["font_size"] = *font_size
		params["color"] = *color
		params["bgcolor"] = *bgcolor
		params["style"] = *style
		params["orientation"] = *orientation
		params["speed"] = *speed
	} else if (*libName) == "datetime" {
		params["zIndex"] = *zIndex
		params["left"] = x
		params["top"] = y
		params["width"] = width
		params["height"] = height
		params["screen_mode"] = *screen_mode
		params["screen_rotate"] = *screen_rotate
		params["font_ttf"] = *font_ttf
		params["font_size"] = *font_size
		params["color"] = *color
		params["bgcolor"] = *bgcolor
		params["align"] = *align
		params["style"] = *style
	} else if (*libName) == "toast" {
		params["zIndex"] = *zIndex
		params["screen_mode"] = *screen_mode
		params["screen_rotate"] = *screen_rotate
		params["content"] = *content
		params["toast_type"] = *toast_type
		params["duration"] = *duration
	}
	data := make(map[string]interface{})
	data["id"] = fmt.Sprintf("PLAY_Z%d_%s_%d", *zIndex, strings.ToUpper(*libName), time.Now().Unix())
	data["type"] = "play"
	data["start"] = *start
	data["libName"] = *libName
	data["params"] = params
	return this.send(data)
}

func (this *XPlay) move() error {
	params := make(map[string]interface{})
	rs := strings.Split(*rect, ",")
	x, _ := strconv.Atoi(rs[0])
	y, _ := strconv.Atoi(rs[1])
	width, _ := strconv.Atoi(rs[2])
	height, _ := strconv.Atoi(rs[3])
	params["zIndex"] = *zIndex
	params["left"] = x
	params["top"] = y
	params["width"] = width
	params["height"] = height
	data := make(map[string]interface{})
	data["type"] = "move"
	data["params"] = params
	return this.send(data)
}

func (this *XPlay) change() error {
	params := make(map[string]interface{})
	params["zIndex"] = *zIndex
	params["newIndex"] = *newIndex
	data := make(map[string]interface{})
	data["type"] = "change"
	data["params"] = params
	return this.send(data)
}

func (this *XPlay) snap() error {
	params := make(map[string]interface{})
	params["path"] = *path
	data := make(map[string]interface{})
	data["type"] = "snap"
	data["params"] = params
	return this.send(data)
}

func (this *XPlay) query() error {
	data := make(map[string]interface{})
	data["type"] = "query"
	return this.send(data)
}

// TODO 暂时只支持视频 ...
func (this *XPlay) path() string {
	// 路径不能为空 ...
	if *path == "" {
		log.Fatal("not found path file ...")
	}
	// 本机播放 ...
	if *addr == "127.0.0.1:8700" {
		// 当前路径文件 ...
		if strings.HasPrefix(*path, "./") {
			dir, _ := os.Getwd()
			ph := dir + string(os.PathSeparator) + (*path)[2:]
			return ph
		}
	} else { // 远端播放 ...
		// 当前路径文件 ...
		if strings.HasPrefix(*path, "./") {
			go func() {
				wg.Add(1) // 添加计数
				// 启动文件服务 ...
				dir, _ := os.Getwd()
				http.Handle("/xplay/", http.StripPrefix("/xplay/", http.FileServer(http.Dir(dir))))
				log.Fatal(http.ListenAndServe(":8711", nil))
				wg.Done() // 移除计数
			}()
			return "http://" + this.ipaddr() + ":8711/xplay/" + (*path)[2:]
		}
	}
	return *path
}

func (this *XPlay) ipaddr() string {
	addrs, err := net.InterfaceAddrs()
	if err != nil {
		log.Fatal(err.Error())
	}
	for _, address := range addrs {
		// 检查IP地址判断是否回环地址
		if ipnet, ok := address.(*net.IPNet); ok && !ipnet.IP.IsLoopback() {
			if ipnet.IP.To4() != nil {
				ip := ipnet.IP.String()
				a1 := strings.Split(*addr, ":")
				if len(a1) == 2 {
					a2 := strings.Split(a1[0], ".")
					if len(a2) == 4 {
						a3 := a2[0] + "." + a2[1] + "." + a2[2]
						if strings.HasPrefix(ip, a3) {
							return ip
						}
					}
				}
			}
		}
	}
	return ""
}

func main() {

	// 设置一下日志的结构
	log.SetFlags(log.Lshortfile | log.Ltime | log.Lmicroseconds)

	// 解析程序参数
	flag.Parse()

	// New XPlay
	xplay := NewXPlay()

	// 连接到 XPlay
	if err := xplay.connect(*addr); err != nil {
		// 连接异常退出 ...
		log.Fatal(err.Error())
	} else {
		// 结束时断开连接
		defer xplay.conn.Close()
	}

	// Error
	var err error

	// 模式
	if *play { // Play
		err = xplay.play()
	} else if *move { // Move
		err = xplay.move()
	} else if *change { // Change
		err = xplay.change()
	} else if *snap { // Snap
		err = xplay.snap()
	} else if *stop { // Stop
		if *all { // Stop ALL
			err = xplay.stop_all()
		} else if *ids != "" { // Stop IDS
			err = xplay.stop(*ids)
		}
	} else if *query { // Query
		xplay.query()
	}

	// Print Error
	if err == nil {
		// 接收返回
		xplay.result()
	} else {
		// 输出异常 ...
		log.Fatal(err.Error())
	}

	// 等待计数结束
	log.Println("正在检查推流状态 ...")
	log.Println("等待推流结束 ...")
	log.Println("退出(Ctrl+C)")
	wg.Wait()
	log.Println("完成")

}

File: example/XplayCtl.java
package xplay;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.net.Socket;
import java.net.UnknownHostException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonElement;
import com.google.gson.JsonParser;

class Rect {

	Rect(int _x, int _y, int _width, int _height) {
		x = _x;
		y = _y;
		width = _width;
		height = _height;
	}

	int x;
	int y;
	int width;
	int height;
}

class Dep {
	String path; // 素材路径
	String type; // 素材类型(video or pic)
	int duration; // 素材持续时间(非视频)
}

public class XplayCtl {

	XplayCtl(String _host, int _port) {
		host = _host;
		port = _port;
	}

	private String host;
	private int port;
	private Socket socket = null;

	private boolean connXplay(String _data) {
		try {
			if (socket == null || !socket.isConnected())
				socket = new Socket(host, port);
			OutputStreamWriter os = new OutputStreamWriter(socket.getOutputStream(), "UTF-8");
			os.write(_data + "\n#End\n");
			os.flush();
			BufferedReader br = new BufferedReader(new InputStreamReader(socket.getInputStream(), "UTF-8"));
			String data = null;
			while ((data = br.readLine()) != null) {
				if (data.contains("#End"))
					break;
				System.out.println(data);
			}
		} catch (UnknownHostException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
			return false;
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
			return false;
		}
		return true;
	}

	private boolean playXplay(Map<String, Object> _data) {
		Gson gson = new GsonBuilder().setPrettyPrinting().create();
		String json = gson.toJson(_data);
		JsonParser jp = new JsonParser();
		JsonElement je = jp.parse(json);
		String prettyJsonString = gson.toJson(je);
		System.out.print(prettyJsonString);
		return this.connXplay(prettyJsonString);
	}

	/*
	 * 序列(视频与图片) 
	 * _content 一组素材，每一个素材为一个 Dep 包含素材路径、类型(libName)、持续时间(非视频)
	 */
	public boolean playSequence(long _start, String _id, int _index, ArrayList<Dep> _content, Rect _rect, String _mode,
			int _rotate) {
		Map<String, Object> data = new HashMap<String, Object>();
		Map<String, Object> params = new HashMap<String, Object>();
		data.put("id", _id);
		data.put("type", "sequence");
		data.put("start", _start);
		data.put("libName", "video");
		params.put("zIndex", _index);
		params.put("left", _rect.x);
		params.put("top", _rect.y);
		params.put("width", _rect.width);
		params.put("height", _rect.height);
		params.put("screen_mode", _mode);
		params.put("screen_rotate", _rotate);
		data.put("params", params);
		data.put("deps", _content);
		return this.playXplay(data);
	}

	/*
	 * 视频或流媒体等 ... 
	 * _start 该素材开始播放的时间精确到毫秒 (视频预加载问题:
	 * 提前500ms-1000ms发送指令xplay可以预加载视频)(也就是_start时间是12:00:00则在11:59:59把指令发送xplay) 
	 * _id 该素材唯一标识符(调试使用) 
	 * _index 该素材所使用的层(0-999)(层数越小越靠前，一般1-9层作为保留层，显示一些提示或者LOGO，通知等信息使用 ...) 
	 * _content 素材存储路径
	 * _rect 素材显示位置与尺寸(素材可以任意缩放拉伸，任意坐标显示) 
	 * _mode 屏幕模式(横屏：landscape、竖屏：portrait)
	 * _rotate 旋转角度(横屏：0，180、竖屏：90，270)
	 */
	public boolean playVideo(long _start, String _id, int _index, String _content, Rect _rect, String _mode,
			int _rotate) {
		Map<String, Object> data = new HashMap<String, Object>();
		Map<String, Object> params = new HashMap<String, Object>();
		data.put("id", _id);
		data.put("type", "play");
		data.put("start", _start);
		data.put("libName", "video");
		params.put("zIndex", _index);
		params.put("path", _content);
		params.put("left", _rect.x);
		params.put("top", _rect.y);
		params.put("width", _rect.width);
		params.put("height", _rect.height);
		params.put("screen_mode", _mode);
		params.put("screen_rotate", _rotate);
		data.put("params", params);
		return this.playXplay(data);
	}

	/*
	 * 图片，参数使用与视频相同
	 */
	public boolean playImage(long _start, String _id, int _index, String _content, Rect _rect, String _mode,
			int _rotate) {
		Map<String, Object> data = new HashMap<String, Object>();
		Map<String, Object> params = new HashMap<String, Object>();
		data.put("id", _id);
		data.put("type", "play");
		data.put("start", _start);
		data.put("libName", "pic");
		params.put("zIndex", _index);
		params.put("path", _content);
		params.put("left", _rect.x);
		params.put("top", _rect.y);
		params.put("width", _rect.width);
		params.put("height", _rect.height);
		params.put("screen_mode", _mode);
		params.put("screen_rotate", _rotate);
		data.put("params", params);
		return this.playXplay(data);
	}

	/*
	 * 二维码 _content 内容为二维码中的内容
	 */
	public boolean playQRCode(long _start, String _id, int _index, String _content, Rect _rect, String _mode,
			int _rotate) {
		Map<String, Object> data = new HashMap<String, Object>();
		Map<String, Object> params = new HashMap<String, Object>();
		data.put("id", _id);
		data.put("type", "play");
		data.put("start", _start);
		data.put("libName", "qrcode");
		params.put("zIndex", _index);
		params.put("content", _content);
		params.put("left", _rect.x);
		params.put("top", _rect.y);
		params.put("width", _rect.width);
		params.put("height", _rect.height);
		params.put("screen_mode", _mode);
		params.put("screen_rotate", _rotate);
		data.put("params", params);
		return this.playXplay(data);
	}

	/*
	 * GIF，参数使用与视频相同
	 */
	public boolean playGIF(long _start, String _id, int _index, String _content, Rect _rect, String _mode,
			int _rotate) {
		Map<String, Object> data = new HashMap<String, Object>();
		Map<String, Object> params = new HashMap<String, Object>();
		data.put("id", _id);
		data.put("type", "play");
		data.put("start", _start);
		data.put("libName", "gif");
		params.put("zIndex", _index);
		params.put("path", _content);
		params.put("left", _rect.x);
		params.put("top", _rect.y);
		params.put("width", _rect.width);
		params.put("height", _rect.height);
		params.put("screen_mode", _mode);
		params.put("screen_rotate", _rotate);
		data.put("params", params);
		return this.playXplay(data);
	}

	/*
	 * 信息提示框 _content 提示框内容 
	 * _toast_type 提示框类型(notice、success、warning、error)
	 * _duration 提示框显示时间(如果为0则永久显示)
	 */
	public boolean playToast(long _start, String _id, int _index, String _content, String _mode, int _rotate,
			String _toast_type, int _duration) {
		Map<String, Object> data = new HashMap<String, Object>();
		Map<String, Object> params = new HashMap<String, Object>();
		data.put("id", _id);
		data.put("type", "play");
		data.put("start", _start);
		data.put("libName", "toast");
		params.put("zIndex", _index);
		params.put("screen_mode", _mode);
		params.put("screen_rotate", _rotate);
		params.put("content", _content);
		params.put("toast_type", _toast_type);
		params.put("duration", _duration);
		data.put("params", params);
		return this.playXplay(data);
	}

	/*
	 * 摄像头
	 *  _camera_width 与 _camera_height 为摄像头采集画面分辨率，建议(1280x720)
	 */
	public boolean playCamera(long _start, String _id, int _index, String _content, Rect _rect, String _mode,
			int _rotate, int _camera_width, int _camera_height) {
		Map<String, Object> data = new HashMap<String, Object>();
		Map<String, Object> params = new HashMap<String, Object>();
		data.put("id", _id);
		data.put("type", "play");
		data.put("start", _start);
		data.put("libName", "camera");
		params.put("zIndex", _index);
		params.put("device", _content);
		params.put("camera_width", _camera_width);
		params.put("camera_height", _camera_height);
		params.put("left", _rect.x);
		params.put("top", _rect.y);
		params.put("width", _rect.width);
		params.put("height", _rect.height);
		params.put("screen_mode", _mode);
		params.put("screen_rotate", _rotate);
		data.put("params", params);
		return this.playXplay(data);
	}

	/*
	 * 文本 
	 * _color "rgba(0, 128, 0, 100%)" 指定文本颜色RGB及透明度A 
	 * _bgcolor rgba(0, 0, 0, 0%) 指定背景颜色RGB及透明度 
	 * _style 文本样式，支持 normal、bold、italic、underline、strikethrough
	 * _align 对齐方式，支持 center、right、left _ptsize 字体大小 _ttf 指定字体文件，空则使用默认字体
	 */
	public boolean playText(long _start, String _id, int _index, String _content, Rect _rect, String _mode, int _rotate,
			String _ttf, int _ptsize, String _color, String _bgcolor, String _style, String _align) {
		Map<String, Object> data = new HashMap<String, Object>();
		Map<String, Object> params = new HashMap<String, Object>();
		data.put("id", _id);
		data.put("type", "play");
		data.put("start", _start);
		data.put("libName", "text");
		params.put("zIndex", _index);
		params.put("left", _rect.x);
		params.put("top", _rect.y);
		params.put("width", _rect.width);
		params.put("height", _rect.height);
		params.put("screen_mode", _mode);
		params.put("screen_rotate", _rotate);
		params.put("content", _content);
		params.put("color", _color);
		params.put("bgcolor", _bgcolor);
		params.put("font_size", _ptsize);
		params.put("align", _align);
		params.put("style", _style);
		data.put("params", params);
		return this.playXplay(data);
	}

	/*
	 * 滚动字幕 滚动字幕只能支持 横向全屏可自定义高 或者 竖向全屏可自定义宽 ... 
	 * _orientation 移动方向 horizontal、vertical 
	 * _speed 移动速度 默认为 1, 也就是每帧移动一个像素 ，如果 FPS30 也就是 _speed = 1 每秒 30个像素
	 */
	public boolean playScroll(long _start, String _id, int _index, String _content, Rect _rect, String _mode,
			int _rotate, String _ttf, int _ptsize, String _color, String _bgcolor, String _style, String _orientation,
			int _speed) {
		Map<String, Object> data = new HashMap<String, Object>();
		Map<String, Object> params = new HashMap<String, Object>();
		data.put("id", _id);
		data.put("type", "play");
		data.put("start", _start);
		data.put("libName", "scroll");
		params.put("zIndex", _index);
		params.put("left", _rect.x);
		params.put("top", _rect.y);
		params.put("width", _rect.width);
		params.put("height", _rect.height);
		params.put("screen_mode", _mode);
		params.put("screen_rotate", _rotate);
		params.put("content", _content);
		params.put("color", _color);
		params.put("bgcolor", _bgcolor);
		params.put("font_size", _ptsize);
		params.put("style", _style);
		params.put("orientation", _orientation);
		params.put("speed", _speed);
		data.put("params", params);
		return this.playXplay(data);
	}

	/*
	 * 日期时间
	 */
	public boolean playDateTime(long _start, String _id, int _index, Rect _rect, String _mode, int _rotate, String _ttf,
			int _ptsize, String _color, String _bgcolor, String _style, String _align) {
		Map<String, Object> data = new HashMap<String, Object>();
		Map<String, Object> params = new HashMap<String, Object>();
		data.put("id", _id);
		data.put("type", "play");
		data.put("start", _start);
		data.put("libName", "datetime");
		params.put("zIndex", _index);
		params.put("left", _rect.x);
		params.put("top", _rect.y);
		params.put("width", _rect.width);
		params.put("height", _rect.height);
		params.put("screen_mode", _mode);
		params.put("screen_rotate", _rotate);
		params.put("color", _color);
		params.put("bgcolor", _bgcolor);
		params.put("font_size", _ptsize);
		params.put("align", _align);
		params.put("style", _style);
		params.put("font_ttf", _ttf);
		data.put("params", params);
		return this.playXplay(data);
	}

	/*
	 * 移动当前层坐标 ...
	 */
	public boolean moveIndex(int _index, int _x, int _y) {
		Map<String, Object> data = new HashMap<String, Object>();
		Map<String, Object> params = new HashMap<String, Object>();
		data.put("type", "move");
		params.put("zIndex", _index);
		params.put("left", _x);
		params.put("top", _y);
		data.put("params", params);
		return this.playXplay(data);
	}

	/*
	 * 停止全部层或者指定层 ...
	 */
	public boolean stopIndex(boolean _all, ArrayList<String> _indexs) {
		Map<String, Object> data = new HashMap<String, Object>();
		Map<String, Object> params = new HashMap<String, Object>();
		data.put("type", "stop");
		if (_all) {
			params.put("all", true);
		} else {
			params.put("ids", _indexs);
		}
		data.put("params", params);
		return this.playXplay(data);
	}

	/*
	 * 截屏 
	 * _path 截屏文件所保存到的路径(例如:/dev/shm/snap.jpg)
	 */
	public boolean snapXplay(String _path) {
		Map<String, Object> data = new HashMap<String, Object>();
		Map<String, Object> params = new HashMap<String, Object>();
		data.put("type", "snap");
		params.put("path", _path);
		data.put("params", params);
		return this.playXplay(data);
	}

}
